### PR TITLE
Update Scan Gibbs sampling example

### LINKED ISSUE #890: Scan Gibbs example is outdated and wrong
This section of the Scan documentation has several issues:

https://aesara.readthedocs.io/en/latest/library/scan.html#using-shared-variables-gibbs-sampling

1. Binomial no longer accepts `n`, `p` as keyword arguments 
2. Outputs info is wrong because it uses a `float` vector, but the returned output type is `int64`
3. `aesara.dot` is no longer a thing
4. updates are not actually optional, the function will not compile without them #579

### PR DESCRIPTION
**Pull request by LegrandNico**
This PR tries to fix #890.

The example is now working with the last version of Aesara.

### CODE REVIEW

File: doc/library/scan.rst [hunk 1]
```diff
@@ -257,32 +257,33 @@ the following:
...
+   import aesara. tensor as at
 
    W = aesara.shared(W_values) # we assume that ``W_values`` contains the
                                # initial values of your weight matrix
 
    bvis = aesara.shared(bvis_values)
    bhid = aesara.shared(bhid_values)
 
-   trng = aesara.tensor.random.utils.RandomStream(1234)
+   trng = at.random.utils.RandomStream(1234)
```
Comment (rlouf) [inline doc/library/scan.rst:276]:
```suggestion
   trng = at.random.RandomStream(1234)
```

Comment (rlouf) [reply doc/library/scan.rst:276]: It's an implicit convention, but we usually use `srng` for `RandomStream`s

File: doc/library/scan.rst [hunk 2]
```diff
@@ -257,32 +257,33 @@ the following:
...
+   import numpy as np
+   W_values = np.random.random((2, 2))
+   bvis_values = np.random.random((2,))
+   bhid_values = np.random.random((2,))
 
 .. testcode:: scan1
 
    import aesara
-   from aesara import tensor as at
+   import aesara. tensor as at
```
Comment (rlouf) [inline doc/library/scan.rst:268]:
```suggestion
   import aesara.tensor as at
```

File: doc/library/scan.rst [hunk 3]
```diff
@@ -257,32 +257,33 @@ the following:
...
+   import aesara.tensor as at
 
    W = aesara.shared(W_values) # we assume that ``W_values`` contains the
                                # initial values of your weight matrix
 
    bvis = aesara.shared(bvis_values)
    bhid = aesara.shared(bhid_values)
 
-   trng = aesara.tensor.random.utils.RandomStream(1234)
+   trng = at.random.RandomStream(1234)
```
Comment (rlouf) [inline doc/library/scan.rst:276]:
```suggestion
   srng = at.random.RandomStream(1234)
```

File: doc/library/scan.rst [hunk 4]
```diff
@@ -257,32 +257,33 @@ the following:
...
+       hsample = srng.binomial(1, hmean, size=hmean.shape)
+       vmean = at.sigmoid(at.dot(hsample, W.T) + bvis)
 
-   def OneStep(vsample) :
-       hmean = at.sigmoid(aesara.dot(vsample, W) + bhid)
-       hsample = trng.binomial(size=hmean.shape, n=1, p=hmean)
-       vmean = at.sigmoid(aesara.dot(hsample, W.T) + bvis)
-       return trng.binomial(size=vsample.shape, n=1, p=vmean,
-                            dtype=aesara.config.floatX)
+       return trng.binomial(1, vmean, size=vsample.shape)
```
Comment (rlouf) [inline doc/library/scan.rst:284]:
```suggestion
       return srng.binomial(1, vmean, size=vsample.shape)
```

**Comment by rlouf** [comment]
Thank you for contributing. I tried to make it run on my machine (with `matrix` and `vector`s instead of shared variables), and the comments are left are the took things that prevented it from running.

File: doc/library/scan.rst [hunk 5]
```diff
@@ -257,32 +257,33 @@ the following:
...
    bhid = aesara.shared(bhid_values)
 
-   trng = aesara.tensor.random.utils.RandomStream(1234)
+   srng = at.random.RandomStream(1234)
+
+   def OneStep(vsample):
+
+       hmean = at.sigmoid(at.dot(vsample, W) + bhid)
+       hsample = srng.binomial(1, hmean, size=hmean.shape)
+       vmean = at.sigmoid(at.dot(hsample, W.T) + bvis)
```
Comment (brandonwillard) [inline doc/library/scan.rst:282]:
```suggestion
   def one_step(vsample):
       hmean = at.sigmoid(at.dot(vsample, W) + bhid)
       hsample = srng.binomial(1, hmean, size=hmean.shape)
       vmean = at.sigmoid(at.dot(hsample, W.T) + bvis)
```

File: doc/library/scan.rst [hunk 6]
```diff
@@ -257,32 +257,33 @@ the following:
...
-       hsample = trng.binomial(size=hmean.shape, n=1, p=hmean)
-       vmean = at.sigmoid(aesara.dot(hsample, W.T) + bvis)
-       return trng.binomial(size=vsample.shape, n=1, p=vmean,
-                            dtype=aesara.config.floatX)
+       return srng.binomial(1, vmean, size=vsample.shape)
 
-   sample = aesara.tensor.vector()
+   sample = aesara.tensor.lvector()
 
    values, updates = aesara.scan(OneStep, outputs_info=sample, n_steps=10)
```
Comment (brandonwillard) [inline doc/library/scan.rst:288]:
```suggestion
   values, updates = aesara.scan(one_step, outputs_info=sample, n_steps=10)
```

File: doc/library/scan.rst [hunk 7]
```diff
@@ -257,32 +257,33 @@ the following:
...
 
    import aesara
-   import numpy
-   W_values = numpy.random.random((2, 2))
-   bvis_values = numpy.random.random((2,))
-   bhid_values = numpy.random.random((2,))
+   import numpy as np
+   W_values = np.random.random((2, 2))
+   bvis_values = np.random.random((2,))
+   bhid_values = np.random.random((2,))
```
Comment (brandonwillard) [inline doc/library/scan.rst:263]:
```suggestion
   import numpy as np
   
   rng = np.random.default_rng(203940)
   W_values = rng.normal(size=(2, 2))
   bvis_values = rng.normal(size=(2,))
   bhid_values = rng.normal(size=(2,))
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Just some formatting changes to match the rest of the library.

N.B. We really should find a way to run black on our code snippets in the documentation.

**Comment by rlouf** [comment]
> N.B. We really should find a way to run `black` on our code snippets in the documentation.

And run the examples in the CI! For instance this one is not going to compile because of the use of `shared` variables (which are not necessary).

**Comment by rlouf** [comment]
Unless I missed something the values of the shared variables were not specified.

**Comment by brandonwillard** [comment]
> Unless I missed something the values of the shared variables were not specified.

They're defined, I believe, just not in the same code block.

**Comment by LegrandNico** [comment]
> > Unless I missed something the values of the shared variables were not specified.
> 
> They're defined, I believe, just not in the same code block.

I think this is making the example a bit confusing as the creation of the input variables is hidden from the user, and then the code is commented to mention you should presume the variables already exist. It would probably be easier for the understanding to have everything in the same block.

**Comment by rlouf** [comment]
> It would probably be easier for the understanding to have everything in the same block.

Indeed I always like to copy/paste examples from the documentation when I start playing with a library. I'm sure I'm not the only one.

**Pull request merged**